pkg/trace/agent/obfuscate.go

```diff
@@ -0,0 +1,178 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package agent
+
+import (
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/obfuscate"
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/traceutil"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	tagRedisRawCommand  = "redis.raw_command"
+	tagMemcachedCommand = "memcached.command"
+	tagMongoDBQuery     = "mongodb.query"
+	tagElasticBody      = "elasticsearch.body"
+	tagSQLQuery         = "sql.query"
+	tagHTTPURL          = "http.url"
+)
+
+const (
+	textNonParsable = "Non-parsable SQL query"
+)
+
+func (a *Agent) obfuscateSpan(span *pb.Span) {
+	o := a.obfuscator
+	switch span.Type {
+	case "sql", "cassandra":
+		if span.Resource == "" {
+			return
+		}
+		oq, err := o.ObfuscateSQLString(span.Resource)
+		if err != nil {
+			// we have an error, discard the SQL to avoid polluting user resources.
+			log.Debugf("Error parsing SQL query: %v. Resource: %q", err, span.Resource)
+			if span.Meta == nil {
+				span.Meta = make(map[string]string, 1)
+			}
+			if _, ok := span.Meta[tagSQLQuery]; !ok {
+				span.Meta[tagSQLQuery] = textNonParsable
+			}
+			span.Resource = textNonParsable
+			return
+		}
+
+		span.Resource = oq.Query
+
+		if len(oq.TablesCSV) > 0 {
+			traceutil.SetMeta(span, "sql.tables", oq.TablesCSV)
+		}
+		if span.Meta != nil && span.Meta[tagSQLQuery] != "" {
+			// "sql.query" tag already set by user, do not change it.
+			return
+		}
+		traceutil.SetMeta(span, tagSQLQuery, oq.Query)
+	case "redis":
+		span.Resource = o.QuantizeRedisString(span.Resource)
+		if a.conf.Obfuscation.Redis.Enabled {
+			if span.Meta == nil || span.Meta[tagRedisRawCommand] == "" {
+				// nothing to do
+				break
+			}
+			span.Meta[tagRedisRawCommand] = o.ObfuscateRedisString(span.Meta[tagRedisRawCommand])
+		}
+	case "memcached":
+		if a.conf.Obfuscation.Memcached.Enabled {
+			v, ok := span.Meta[tagMemcachedCommand]
+			if span.Meta == nil || !ok {
+				break
+			}
+			span.Meta[tagMemcachedCommand] = o.ObfuscateMemcachedString(v)
+		}
+	case "web", "http":
+		if span.Meta == nil {
```

Comment:
why do we sometimes check for nil Meta, and sometimes not?

Reply:
I think each branch does. Take a closer look.